src/graph/ChcGraph.cc

```diff
@@ -599,4 +601,126 @@ std::unique_ptr<ChcDirectedHyperGraph> ChcDirectedHyperGraph::makeEmpty(Logic &
         std::move(predicateRepresentation),
         logic
         );
+}
+
+WitnessInfo ChcDirectedGraph::contractConnectedVertexes(std::vector<EId> edges) {
+    assert(edges.size() >= 2);
+    assert(getSource(edges[0]) == getTarget(edges[edges.size() - 1]));
+    TimeMachine timeMachine(logic);
+    TermUtils utils(logic);
+    std::vector<SymRef> vertices;
+    assert(std::adjacent_find(edges.begin(), edges.end(), [&](auto first, auto second) { return getTarget(first) != getSource(second); }) == edges.end());
+    for (int i = edges.size() - 1; i >= 0; i--) {
+        vertices.push_back(getSource(edges[i]));
+    }
+
+    LocationVarMap locationVars;
+    PositionVarMap argVars;
+    locationVars.reserve(vertices.size());
+    for (auto vertex : vertices) {
+        auto varName = std::string(".loc_") + std::to_string(vertex.x);
+        locationVars.insert({vertex, timeMachine.getVarVersionZero(varName, logic.getSort_bool())});
+    }
+
+    for (auto vertex : vertices) {
+        auto args_count = logic.getSym(vertex).nargs();
+        for (uint32_t i = 0; i < args_count; ++i) {
+            VarPosition pos = {vertex, i};
+            auto varName = std::string(".arg_") + std::to_string(vertex.x) + '_' + std::to_string(i);
+            PTRef var = timeMachine.getVarVersionZero(varName, logic.getSym(vertex)[i]);
+            argVars.insert({pos, var});
+        }
+    }
+
+    vec<PTRef> transitionRelationComponent;
+
+    // KB: creation of the location variables for nodes and argument variables (params of predicates)
+    std::vector<EId> allEdges = getEdges();
+    EdgeTranslator edgeTranslator{*this, locationVars, argVars, {}};
+    // KB: Translation of self-loop edges for the edges vertices
+    //     and edges between vertices in DAG
+    for (auto edge : allEdges) {
+        if (std::find(vertices.begin(), vertices.end(), getSource(edge)) != vertices.end() &&
+            std::find(vertices.begin(), vertices.end(), getTarget(edge)) != vertices.end()) {
+            transitionRelationComponent.push(edgeTranslator.translateEdge(getEdge(edge)));
+            removeEdge(edge);
+        }
+    };
+
+    // KB: Extraction of the state variables of the nested loop nodes, creation of new predicate to replace node with
+    //     nested loops
+    std::vector<PTRef> stateVars = [&locationVars, &argVars]() {
+        std::vector<PTRef> ret;
+        for (auto && entry : locationVars) {
+            ret.push_back(entry.second);
+        }
+        for (auto && entry : argVars) {
+            ret.push_back(entry.second);
+        }
+        return ret;
+    }();
+
+    vec<SRef> argSorts;
+    std::vector<PTRef> vars;
+    for (auto arg : stateVars) {
+        argSorts.push(logic.getSortRef(arg));
+        vars.push_back(timeMachine.getUnversioned(arg));
+    }
+
+    PTRef transitionRelation = logic.mkOr(std::move(transitionRelationComponent));
```

Comment:
Move up just after you compute all the components.

Reply:
Fixed